### PR TITLE
api: support setting trusted CIDRs

### PR DESCRIPTION
**Pull request by rudrakhp**
**What type of PR is this?**

api: support setting trusted CIDRs

**What this PR does / why we need it**:

To configure Envoy API's [xff_trusted_cidrs](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto#L48) param.

**Which issue(s) this PR fixes**:

Ref #4489 

Release Notes: Yes

### CODE REVIEW

**Comment by rudrakhp** [comment]
Wanted to add docs after implementation is done but looks like a recent gen check is not allowing me to do that.

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 1]
```diff
@@ -237,14 +237,25 @@ type ClientIPDetectionSettings struct {
...
 
 // XForwardedForSettings provides configuration for using X-Forwarded-For headers for determining the client IP address.
+// Refer to https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for
+// for more details.
 type XForwardedForSettings struct {
 	// NumTrustedHops controls the number of additional ingress proxy hops from the right side of XFF HTTP
 	// headers to trust when determining the origin client's IP address.
-	// Refer to https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for
-	// for more details.
+	// Only one of NumTrustedHops and TrustedCIDRs can be set.
```
Comment (zhaohuabing) [inline api/v1alpha1/clienttrafficpolicy_types.go:245]:
```suggestion
	// Only one of NumTrustedHops or TrustedCIDRs must be set
```

Could you please also add a CEL validation and test for this? You should find many examples in the exiting APIs.

Comment (rudrakhp) [reply api/v1alpha1/clienttrafficpolicy_types.go]: @zhaohuabing done!

File: release-notes/current.yaml [hunk 2]
```diff
@@ -1,29 +1,5 @@
...
-  - Add a deprecation here
-
-# Other notable changes not covered by the above sections.
-Other changes: |
-  - Add other changes here
+date: Oct 23, 2022
+changes:
+  - area: api
+    change: |
+      Support setting trusted CIDRs
```
Comment (zhaohuabing) [inline release-notes/current.yaml:5]: We're adopting a new release format. This change can be added to the new features section.

File: release-notes/current.yaml [hunk 3]
```diff
@@ -1,29 +1,5 @@
...
-performance improvements: |
-  - Add a performance improvement here
-
-# Deprecated features or APIs.
-deprecations: |
-  - Add a deprecation here
-
-# Other notable changes not covered by the above sections.
-Other changes: |
-  - Add other changes here
```
Comment (zhaohuabing) [inline release-notes/current.yaml:29]: Don't delete the original content in this template.

File: site/content/en/news/releases/notes/current.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
```
[the remark this answers is not in the document]
Comment (rudrakhp) [reply site/content/en/news/releases/notes/current.md:1]: @zhaohuabing this is being modified by `gen-check` somehow. Let me retry.

[the remark this answers is not in the document]
Comment (zhaohuabing) [reply site/content/en/news/releases/notes/current.md:1]: @rudrakhp  This has been fixed by https://github.com/envoyproxy/gateway/pull/4521. Merging the main branch will fix it.

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 5]
```diff
@@ -237,14 +237,26 @@ type ClientIPDetectionSettings struct {
 }
 
 // XForwardedForSettings provides configuration for using X-Forwarded-For headers for determining the client IP address.
+// Refer to https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for
+// for more details.
+// +kubebuilder:validation:XValidation:rule="(has(self.numTrustedHops) && !has(self.trustedCIDRs)) || (!has(self.numTrustedHops) && has(self.trustedCIDRs))", message="only one of numTrustedHops or trustedCIDRs must be set"
```
Comment (zhaohuabing) [inline api/v1alpha1/clienttrafficpolicy_types.go:242]: Can we also add tests to https://github.com/envoyproxy/gateway/blob/main/test/cel-validation/clienttrafficpolicy_test.go to make sure the CEL works?

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 6]
```diff
@@ -237,14 +237,26 @@ type ClientIPDetectionSettings struct {
...
 	// +optional
 	NumTrustedHops *uint32 `json:"numTrustedHops,omitempty"`
+
+	// TrustedCIDRs is a list of trusted CIDRs to trust when
+	// evaluating the remote IP address to determine the original client's IP address.
+	// Only one of NumTrustedHops and TrustedCIDRs must be set.
+	//
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:ItemsFormat=cidr
```
Comment (zhaohuabing) [inline api/v1alpha1/clienttrafficpolicy_types.go:257]: kubebuilder support this?

File: release-notes/current.yaml [hunk 7]
```diff
@@ -10,7 +10,7 @@ security updates: |
 
 # New features or capabilities added in this release.
 new features: |
-  - Add a new feature here
+  - api: support setting trusted CIDRs
```
Comment (zhaohuabing) [inline release-notes/current.yaml:13]:
```suggestion
  - Added support for trusted CIDRs in the ClientIPDetectionSettings API
```

File: release-notes/current.yaml [hunk 8]
```diff
@@ -1,4 +1,4 @@
-date: Pending
+date: October 24, 2024
```
Comment (zhaohuabing) [inline release-notes/current.yaml:1]: Please leave the date as Pending.

File: api/v1alpha1/clienttrafficpolicy_types.go (hunk 6 above)
Comment (rudrakhp) [reply api/v1alpha1/clienttrafficpolicy_types.go]: Added regex for items instead. Added a test to validate it.

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 9]
```diff
@@ -237,14 +237,26 @@ type ClientIPDetectionSettings struct {
...
 	// NumTrustedHops controls the number of additional ingress proxy hops from the right side of XFF HTTP
 	// headers to trust when determining the origin client's IP address.
-	// Refer to https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for
-	// for more details.
+	// Only one of NumTrustedHops and TrustedCIDRs must be set.
 	//
 	// +optional
 	NumTrustedHops *uint32 `json:"numTrustedHops,omitempty"`
+
+	// TrustedCIDRs is a list of trusted IPv4 CIDRs to trust when
```
Comment (rudrakhp) [inline api/v1alpha1/clienttrafficpolicy_types.go:251]: @zhaohuabing to confirm - I am assuming gateway has support IPv4 only.

Comment (zirain) [reply api/v1alpha1/clienttrafficpolicy_types.go:251]: IPv6 will be supported in `v1.2`.

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 10]
```diff
@@ -237,14 +237,26 @@ type ClientIPDetectionSettings struct {
...
+	// evaluating the remote IP address to determine the original client's IP address.
+	// Only one of NumTrustedHops and TrustedCIDRs must be set.
+	//
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:items:Pattern=`^((25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9]?[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9]?[0-9])/(3[0-2]|[12]?[0-9])$`
+	// +notImplementedHide
+	TrustedCIDRs []string `json:"trustedCIDRs,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/clienttrafficpolicy_types.go:259]: We can resuse the `CIDR` type.
```suggestion
	TrustedCIDRs []CIDR `json:"trustedCIDRs,omitempty"`
```

File: api/v1alpha1/clienttrafficpolicy_types.go (hunk 6 above)
Comment (zhaohuabing) [reply api/v1alpha1/clienttrafficpolicy_types.go]: We can reuse the existing `CIDR` type.

File: api/v1alpha1/clienttrafficpolicy_types.go (hunk 10 above)
Comment (rudrakhp) [reply api/v1alpha1/clienttrafficpolicy_types.go]: Thanks for this, one less regex to maintain 😌

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 11]
```diff
@@ -237,14 +237,25 @@ type ClientIPDetectionSettings struct {
...
 	// headers to trust when determining the origin client's IP address.
-	// Refer to https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for
-	// for more details.
+	// Only one of NumTrustedHops and TrustedCIDRs must be set.
 	//
 	// +optional
 	NumTrustedHops *uint32 `json:"numTrustedHops,omitempty"`
+
+	// TrustedCIDRs is a list of CIDRs to trust when evaluating
+	// the remote IP address to determine the original client's IP address.
```
Comment (zhaohuabing) [inline api/v1alpha1/clienttrafficpolicy_types.go:252]: **Nit, non-blocking - the explaination of TrustedCIDRs is a bit vague here. Suggest add a clearer explaination like the blow:
```suggestion
	// TrustedCIDRs is a list of CIDR ranges to trust when evaluating the remote IP address to determine the original client’s IP address. 
	// When the remote IP address matches a trusted CIDR and the x-forwarded-for header was sent, each entry in the x-forwarded-for header is evaluated from right to left and the first public non-trusted address is used as the original client address. If all addresses in x-forwarded-for are within the trusted list, the first (leftmost) entry is used.
	// the remote IP address to determine the original client's IP address.
```**

**Review by zhaohuabing — DISMISSED** [review]
A nit on the comments, it's non-blocking.

File: api/v1alpha1/clienttrafficpolicy_types.go (hunk 11 above)
Comment (rudrakhp) [reply api/v1alpha1/clienttrafficpolicy_types.go]: Done!

**Review by zhaohuabing — APPROVED** [review]
The API LGTM, thanks!

**Comment by arkodg** [comment]
Let's hold off on merging API and feature PRs until v1.2 is released to simplify cherry picking

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Pull request merged**